docs/codeql/codeql-cli/about-codeql-workspaces.rst

```diff
@@ -0,0 +1,84 @@
+
+For example, the following ``codeql-workspace.yml`` file defines a workspace that contains all CodeQL packs recursively found in the ``codeql-packs`` directory, except for the packs in the ``experimental`` directory:
+
+.. code-block:: yaml
+
+    provide:
+      - "*/codeql-packs/**/qlpack.yml"
+    ignore:
+      - "*/codeql-packs/**/experimental/**/qlpack.yml"
+
+To verify that you have the correct ``codeql-workspace.yml`` file, run ``codeql pack ls`` command in the same directory as your workspace. The result of the command is a list of all CodeQL packs in the workspace.
+
+
+CodeQL workspaces and query resolution
+--------------------------------------
+
+All CodeQL packs in a workspace are available as source dependencies for each other when running any CodeQL command that resolves queries or packs. For example, when ``codeql pack install`` is run in a pack directory in a workspace, any dependency found in the workspace will not be downloaded to the package cache, nor will it be added to the resulting ``codeql-pack.lock.yml`` file. See `:ref:Adding and Installing Dependencies <adding-and-installing-dependencies>` for more information.
+
+Similarly, publishing a CodeQL query pack to the GitHub container registry using  ``codeql pack publish`` will always use dependencies found in the workspace instead of using dependencies found in the local package cache.
+
+This ensures that any local change to a query library in a dependency in the same workspace will be automatically reflected in the published query pack.
+
+.. pull-quote::
+
+  Note
+
+  Source dependencies are CodeQL packs that are resolved from the filesystem. They might be in the same CodeQL workspace, or specified a path option in the ``--additional-packs`` argument. Source dependencies override any dependencies found in the local package cache and version constraints are ignored. This ensures that during local development version mismatches can be ignored.
+
+Example
+~~~~~~~
+
+Consider the following ``codeql-workspace.yml`` file:
+
+.. code-block:: yaml
+
+    provide:
+      - "**/qlpack.yml"
+
+And the following CodeQL library pack ``qlpack.yml`` file in the workspace:
+
+.. code-block:: yaml
+
+    name: my-company/my-library
+    library: true
+    version: 1.0.0
+
+And the following CodeQL query pack ``qlpack.yml`` file in the workspace:
+
+.. code-block:: yaml
+
+    name: my-company/my-queries
+    version: 1.0.0
+    dependencies:
+      my-company/my-library: "*"
+      codeql/cpp-all: ~0.2.0
+
+Notice that ``"*"`` is specified as the version constraint for the library pack. Because the library pack is a source dependency, the version constraint is not needed since the library pack's content is always resolved from inside of the workspace. Any version constraint will be ignored in this case, but it is recommended to use ``"*"`` for source dependencies to avoid confusion.
```

Comment:
```suggestion
Notice that, for ``my-company/my-queries``, ``"*"`` is specified as the version constraint for the library pack in the ``dependencies`` block. The library pack is defined as a source dependency in ``codeql-workspace.yml``, so the version constraint is not needed since the library pack's content is always resolved from inside of the workspace. Any version constraint will be ignored in this case, but it is recommended to use ``"*"`` for source dependencies to avoid confusion.
```